### PR TITLE
Bring mediacapture-streams tests to spec

### PR DESCRIPTION
**Pull request by Pehrsons**
- Primarily moves navigator.getUserMedia (w/ prefix) to
navigator.mediaDevices.getUserMedia.
- Also brings other items to spec as found while running these tests in Firefox

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248543476)

# Firefox (nightly) #
Testing web-platform-tests at revision 663a867e045bd1366ad54b5969887683657ce217
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>20 tests ran</summary>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-api.https.html">/mediacapture-streams/GUM-api.https.html</a></summary>


|                        Subtest                        | Results | Messages |
|-------------------------------------------------------|---------|----------|
|                                                       | OK      |          |
| `mediaDevices.getUserMedia() is present on navigator` | PASS    |          |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-deny.https.html">/mediacapture-streams/GUM-deny.https.html</a></summary>


|                                Subtest                                 | Results |                                                 Messages                                                 |
|------------------------------------------------------------------------|---------|----------------------------------------------------------------------------------------------------------|
|                                                                        | OK      |                                                                                                          |
| `Tests that the error callback is triggered when permission is denied` | FAIL    | `assert_equals: NotAllowedError returned as expected expected "NotAllowedError" but got "NotFoundError"` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-empty-option-param.https.html">/mediacapture-streams/GUM-empty-option-param.https.html</a></summary>


|                                             Subtest                                              | Results | Messages |
|--------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                  | OK      |          |
| `Tests that getUserMedia is rejected with a TypeError when used with an empty options parameter` | PASS    |          |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-impossible-constraint.https.html">/mediacapture-streams/GUM-impossible-constraint.https.html</a></summary>


|                               Subtest                               | Results |                                                              Messages                                                             |
|---------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------------------------|
|                                                                     | OK      |                                                                                                                                   |
| `Tests that setting an impossible constraint in getUserMedia fails` | FAIL    | `assert_equals: An impossible constraint triggers a OverconstrainedError expected "OverconstrainedError" but got "NotFoundError"` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-optional-constraint.https.html">/mediacapture-streams/GUM-optional-constraint.https.html</a></summary>


|                                      Subtest                                       | Results |                                                     Messages                                                    |
|------------------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------------|
|                                                                                    | OK      |                                                                                                                 |
| `Tests that setting an optional constraint in getUserMedia is handled as optional` | FAIL    | `assert_unreached: an optional constraint can't stop us from obtaining a video stream Reached unreachable code` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-trivial-constraint.https.html">/mediacapture-streams/GUM-trivial-constraint.https.html</a></summary>


|                                  Subtest                                  | Results |                                                  Messages                                                 |
|---------------------------------------------------------------------------|---------|-----------------------------------------------------------------------------------------------------------|
|                                                                           | OK      |                                                                                                           |
| `Tests that setting a trivial mandatory constraint in getUserMedia works` | FAIL    | `assert_unreached: a Video stream of minimally zero width can always be created Reached unreachable code` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/GUM-unknownkey-option-param.https.html">/mediacapture-streams/GUM-unknownkey-option-param.https.html</a></summary>


|                                           Subtest                                           | Results | Messages |
|---------------------------------------------------------------------------------------------|---------|----------|
|                                                                                             | OK      |          |
| `Tests that getUserMedia is rejected with a TypeError when used with an unknown constraint` | PASS    |          |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-MediaElement-preload-none.https.html">/mediacapture-streams/MediaStream-MediaElement-preload-none.https.html</a></summary>


|                                   Subtest                                    | Results |                                        Messages                                       |
|------------------------------------------------------------------------------|---------|---------------------------------------------------------------------------------------|
|                                                                              | OK      |                                                                                       |
| `Test that preload 'none' is ignored for MediaStream object URL used as src` | FAIL    | `assert_unreached: getUserMedia error callback was invoked. Reached unreachable code` |
| `Test that preload 'none' is ignored for MediaStream used as srcObject`      | FAIL    | `assert_unreached: getUserMedia error callback was invoked. Reached unreachable code` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-MediaElement-srcObject.https.html">/mediacapture-streams/MediaStream-MediaElement-srcObject.https.html</a></summary>


|                                   Subtest                                    | Results |     Messages     |
|------------------------------------------------------------------------------|---------|------------------|
|                                                                              | TIMEOUT |                  |
| `Tests that a MediaStream can be assigned to a video element with srcObject` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-add-audio-track.https.html">/mediacapture-streams/MediaStream-add-audio-track.https.html</a></summary>


|                            Subtest                             | Results |     Messages     |
|----------------------------------------------------------------|---------|------------------|
|                                                                | TIMEOUT |                  |
| `Tests that adding a track to a MediaStream works as expected` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-audio-only.https.html">/mediacapture-streams/MediaStream-audio-only.https.html</a></summary>


|                               Subtest                               | Results |     Messages     |
|---------------------------------------------------------------------|---------|------------------|
|                                                                     | TIMEOUT |                  |
| `Tests that a MediaStream with exactly one audio track is returned` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-finished-add.https.html">/mediacapture-streams/MediaStream-finished-add.https.html</a></summary>


|                              Subtest                              | Results |     Messages     |
|-------------------------------------------------------------------|---------|------------------|
|                                                                   | TIMEOUT |                  |
| `Tests that adding a track to an inactive MediaStream is allowed` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-gettrackid.https.html">/mediacapture-streams/MediaStream-gettrackid.https.html</a></summary>


|                         Subtest                         | Results |                                                           Messages                                                           |
|---------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------|
|                                                         | OK      |                                                                                                                              |
| `Tests that MediaStream.getTrackById works as expected` | FAIL    | `assert_unreached: Unexpected getUserMedia error: NotFoundError: The object can not be found here. Reached unreachable code` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-idl.https.html">/mediacapture-streams/MediaStream-idl.https.html</a></summary>


|                                   Subtest                                    | Results |     Messages     |
|------------------------------------------------------------------------------|---------|------------------|
|                                                                              | TIMEOUT |                  |
| `Tests that a MediaStream constructor follows the algorithm set in the spec` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-removetrack.https.html">/mediacapture-streams/MediaStream-removetrack.https.html</a></summary>


|                           Subtest                           | Results |     Messages     |
|-------------------------------------------------------------|---------|------------------|
|                                                             | TIMEOUT |                  |
| `Tests that a removal from a MediaStream works as expected` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStream-video-only.https.html">/mediacapture-streams/MediaStream-video-only.https.html</a></summary>


|                               Subtest                                | Results |     Messages     |
|----------------------------------------------------------------------|---------|------------------|
|                                                                      | TIMEOUT |                  |
| `Tests that a MediaStream with at least one video track is returned` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStreamTrack-MediaElement-disabled-audio-is-silence.https.html">/mediacapture-streams/MediaStreamTrack-MediaElement-disabled-audio-is-silence.https.html</a></summary>


|                                   Subtest                                   | Results |     Messages     |
|-----------------------------------------------------------------------------|---------|------------------|
|                                                                             | TIMEOUT |                  |
| `Tests that a disabled audio track in a MediaStream is rendered as silence` | TIMEOUT | `Test timed out` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStreamTrack-MediaElement-disabled-video-is-black.https.html">/mediacapture-streams/MediaStreamTrack-MediaElement-disabled-video-is-black.https.html</a></summary>


|                      Subtest                      | Results |                  Messages                 |
|---------------------------------------------------|---------|-------------------------------------------|
|                                                   | OK      |                                           |
| `A disabled video track is rendered as blackness` | FAIL    | `SyntaxError: missing ; before statement` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStreamTrack-id.https.html">/mediacapture-streams/MediaStreamTrack-id.https.html</a></summary>


|                           Subtest                           | Results |                Messages                |
|-------------------------------------------------------------|---------|----------------------------------------|
|                                                             | OK      |                                        |
| `Tests that distinct mediastream tracks have distinct ids ` | FAIL    | `t.assert_unreached is not a function` |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/6429/mediacapture-streams/MediaStreamTrack-init.https.html">/mediacapture-streams/MediaStreamTrack-init.https.html</a></summary>


|                                 Subtest                                  | Results |     Messages     |
|--------------------------------------------------------------------------|---------|------------------|
|                                                                          | TIMEOUT |                  |
| `Tests that the video MediaStreamTrack objects are properly initialized` | TIMEOUT | `Test timed out` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248543477)

# Sauce (safari) #
Testing web-platform-tests at revision 663a867e045bd1366ad54b5969887683657ce217
Using browser at version 10.0
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248543485)

# Chrome (unstable) #
Testing web-platform-tests at revision 663a867e045bd1366ad54b5969887683657ce217
Using browser at version 61.0.3141.7 dev
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/248543486)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 663a867e045bd1366ad54b5969887683657ce217
Using browser at version 14.14393
Starting 10 test iterations
No tests run.

**Comment by sideshowbarker** [comment]
w3c-test:mirror

**Comment by jan-ivar** [comment]
@dontcallmedom This appears stuck. Any tips on how to proceed?

**Comment by jan-ivar** [comment]
@soareschen Any chance you could look at these as well? They've been stuck for a bit, and I'm not sure who is actively reviewing things here still.

File: mediacapture-streams/MediaStream-idl.https.html [hunk 1]
```diff
@@ -19,33 +19,33 @@ <h1 class="instructions">Description</h1>
 <div id='log'></div>
 <script src=/resources/testharness.js></script>
 <script src=/resources/testharnessreport.js></script>
-<script src="/common/vendor-prefix.js" data-prefixed-objects='[{"ancestors":["navigator"], "name":"getUserMedia"},{"ancestors":["window"], "name":"MediaStream"}]'></script>
+<script src="/common/vendor-prefix.js" data-prefixed-objects='[{"ancestors":["window"], "name":"MediaStream"}]'></script>
```
Comment (soareschen) [inline mediacapture-streams/MediaStream-idl.https.html]: Is there a reason `MediaStream` remain prefixed? If not it can be removed in a separate PR.

File: mediacapture-streams/MediaStream-idl.https.html [hunk 2]
```diff
@@ -19,33 +19,33 @@ <h1 class="instructions">Description</h1>
...
 
-    var audioTrack2 = audioTrack1.clone();
-    audioTrack2.addEventListener("ended", t.step_func(function () {
+      var audioTrack2 = audioTrack1.clone();
+      audioTrack2.stop();
+      assert_equals(audioTrack2.readyState, "ended", "a stopped track is marked ended synchronously");
       var stream3 = new MediaStream([audioTrack2, videoTrack]);
-      assert_equals(stream3.getTrackById(audioTrack2.id), null, "an ended track doesn't get added via the MediaStream constructor");
-      assert_equals(stream3.getTrackById(videoTrack.id), videoTrack, "a non-ended track gets added via the MediaStream constructor even if the previous track was ended");
+      assert_equals(stream3.getTrackById(audioTrack2.id), audioTrack2, "an ended track gets added via the MediaStream constructor");
```
Comment (soareschen) [inline mediacapture-streams/MediaStream-idl.https.html]: Just to verify, does this correspond to changes in the mediacapture-main spec? I checked the spec and the new code tests the correct behavior.

File: mediacapture-streams/MediaStream-idl.https.html [hunk 3]
```diff
@@ -19,33 +19,33 @@ <h1 class="instructions">Description</h1>
...
+      assert_not_equals(stream.id, stream2.id, "A MediaStream constructed from another have different ids");
+      var audioTrack1 = stream.getAudioTracks()[0];
+      var videoTrack = stream.getVideoTracks()[0];
+      assert_equals(audioTrack1, stream2.getAudioTracks()[0], "A MediaStream constructed from another share the same audio track");
+      assert_equals(videoTrack, stream2.getVideoTracks()[0], "A MediaStream constructed from another share the same video track");
+      var stream4 = new MediaStream([audioTrack1]);
+      assert_equals(stream4.getTrackById(audioTrack1.id), audioTrack1, "a non-ended track gets added via the MediaStream constructor");
 
-    var audioTrack2 = audioTrack1.clone();
-    audioTrack2.addEventListener("ended", t.step_func(function () {
```
Comment (soareschen) [inline mediacapture-streams/MediaStream-idl.https.html]: There is some conflict here with master caused by 530273179c79.

**Review by soareschen — CHANGES_REQUESTED** [review]
Overall I find this ok, other than the merge conflict that need to be resolved and comment below.

Personally I prefer using `promise_test` rather than `async_test` with this change to promise-based API. There are also some errors ignored in the async test cases. But given this is mostly style changes to the original code, I think we can leave those to other PRs for refactoring.

One issue I'd like to follow up is, what about testing the legacy `navigator.getUserMedia()` API? This effectively removes the tests for the legacy API, but they are still in the spec. I notice Firefox have removed `navigator.getUserMedia()`, and thus fail on the original tests. 

I found issue w3c/mediacapture-main#368 which discussed about removing `navigator.getUserMedia()` from the spec, but the conclusion is to keep it in the spec. This means we have to keep some test for the legacy API, and signifies that Firefox need to reimplement them.

File: mediacapture-streams/MediaStream-idl.https.html (hunk 1 above)
Comment (jan-ivar) [reply mediacapture-streams/MediaStream-idl.https.html]: No reason I know of. `MediaStream` hasn't been prefixed in Firefox for ages.

However, `navigator.getUserMedia` still is. @Pehrsons how come we removed it here?

File: mediacapture-streams/MediaStream-idl.https.html (hunk 2 above)
Comment (jan-ivar) [reply mediacapture-streams/MediaStream-idl.https.html]: I don't recall it ever working that way, but I may have missed it.

**Comment by jan-ivar** [comment]
@soareschen Aren't legacy interfaces optional to implement like they are in WebRTC?

Neither Safari nor Firefox implement `navigator.getUserMedia`.

**Comment by soareschen** [comment]
> Aren't legacy interfaces optional to implement like they are in WebRTC?

Ah I see. Missed the sentence _"Supporting the methods in this section is optional"_ in 4.4.3. This PR can be merged after conflict is resolved then. :)

**Review by soareschen — APPROVED** [review]

File: mediacapture-streams/MediaStream-idl.https.html (hunk 1 above)
Comment (Pehrsons) [reply mediacapture-streams/MediaStream-idl.https.html]: I removed `navigator.getUserMedia` from this test in favor of `navigator.mediaDevices.getUserMedia`.

I can remove the MediaStream vendor-prefix here. Just an oversight by me, I grepped for vendor-prefixed getUserMedia and removed those.

**Review by jan-ivar — APPROVED** [review]

**Comment by Pehrsons** [comment]
@soareschen this should be clear to land now. I removed the MediaStream vendor-prefix as mentioned, and resolved the conflict.

Thanks!

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/18398)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/274096242)

File: mediacapture-streams/MediaStream-audio-only.https.html [hunk 4]
```diff
@@ -15,12 +15,12 @@ <h1 class="instructions">Description</h1>
 <div id='log'></div>
 <script src=/resources/testharness.js></script>
 <script src=/resources/testharnessreport.js></script>
-<script src="/common/vendor-prefix.js" data-prefixed-objects='[{"ancestors":["navigator"], "name":"getUserMedia"}, {"ancestors":["window"], "name":"MediaStream"}]'></script>
+<script src="/common/vendor-prefix.js" data-prefixed-objects='[{"ancestors":["window"], "name":"MediaStream"}]'></script>
```
Comment (kaixinjxq) [inline mediacapture-streams/MediaStream-audio-only.https.html]: Why not remove this vendor?

Comment (Pehrsons) [reply mediacapture-streams/MediaStream-audio-only.https.html]: I did a search for getUserMedia and removed those, but sure, I can remove all the MediaStream vendor prefixes too.
